### PR TITLE
  fix volume driver deletions prevents recreation of volume issue.

### PR DESCRIPTION
**Pull request by moypray**
**\- What I did**
fix a issue which volume driver deletions will cause recreation of volume failed.

**\- How I did it**

When creating a volume in docker, it will check the memory cache in docker, I just step over it and use "VolumeDriver.Get" API to check the volume status.

**\- How to verify it**
1.  docker volume create --name test -d convoy -o size=1
2. convoy delete test
3. docker volume ls   #  to check it does not exist.
4. docker volume create --name test -d convoy -o size=1  

**\- Description for the changelog**

**\- A picture of a cute animal (not mandatory but encouraged)**

  Signed-off-by:  Wentao Zhang zhangwentao234@huawei.com

### CODE REVIEW

**Comment by hqhq** [comment]
We might have another issue in scenario like:
- docker volume create --name test -d convoy -o size=1
- convoy delete test
- docker volume ls # to check it does not exist.
- docker volume create --name test -d driver2 -o size=1

I think we should handle the cache issue instead of skip it, otherwise the logic would be ugly.

**Comment by erikh** [comment]
nice. concise. I love it.

**Comment by moypray** [comment]
@erikh   see #23367

The same to you,  I think the memory cache is useless too.  When there is no GET API, maybe we need the cache. But now we have GET API. I think should delete the memory cache.

**Comment by yongtang** [comment]
@moypray The cache of the volume consists of `labels` and `refs`, in addition to `names`. I think you may need to reset the refs as it is only appending to the end when `setNamed` is called.

**Comment by yongtang** [comment]
I may have some concerns overall as well. Like @hqhq mentioned, automatically recreate may cause some unexpected side effects.

File: volume/store/store.go [hunk 1]
```diff
@@ -260,7 +260,10 @@ func (s *VolumeStore) create(name, driverName string, opts, labels map[string]st
 		if v.DriverName() != driverName && driverName != "" && driverName != volume.DefaultDriverName {
 			return nil, errNameConflict
 		}
-		return v, nil
+		// check exist in driver
```
Comment (thaJeztah) [inline volume/store/store.go]: Shouldn't this be above the "if" block above (line 260)? And the `driverName != .....` can be removed from that if (if I read correctly)

**Comment by moypray** [comment]
@yongtang @hqhq   Do it allow to have same name upon different drivers. I have tested that create a volume named "abc" with convoy, and then create a volume named abc too using local driver. but I couldn't see it using the command "docker volume ls"  (1.11.1 docker version)

root@main-server:~/work/elara# docker version
Client:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.2
 Git commit:   e9b27e2-unsupported
 Built:        Fri Jul  1 14:42:18 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.1
 API version:  1.23
 Go version:   go1.5.2
 Git commit:   e70cf10
 Built:        Thu Jun 30 15:48:30 2016
 OS/Arch:      linux/amd64

**Comment by hqhq** [comment]
According to the logic in `create` function:

```
259         if v, exists := s.getNamed(name); exists {
260                 if v.DriverName() != driverName && driverName != "" && driverName != volume.DefaultDriverName {
261                         return nil, errNameConflict
262                 }
263                 return v, nil
264         }
```

Identical volume is not allowed but not for local driver, I surely have missed some background stories but this logic still looks weird to me.

Anyway the issue I pointed out is for different plugin drivers and the original volume is deleted before the same-named-volume is created by another plugin driver, this should be a valid case but it will fail because the cache is not synced.

So the point is there might be more issues duo to the dirty cache, skipping cache might be easy but not the best solution, we need to fix the dirty cache issue, either by syncing it or totally remove the cache.

**Comment by erikh** [comment]
I agree with that. I just am pretty sure this will solve the immediate case.

**Comment by yongtang** [comment]
@moypray @hqhq @erikh @thaJeztah A pull request #23436 that cleans out-of-sync cache was created earlier. Please let me know if there are any feedbacks.

**Comment by moypray** [comment]
@yongtang Got it , if we need to save refs in docker, we shouldn't delete the cache. But I think solve the issue immediately while creating volume will be a better way  as @erikh said.  If so, I will update my PR.

How do you feel ? @hqhq @thaJeztah

**Comment by erikh** [comment]
Please ensure that is the correct solution; I am reading it but my docker
dev environment hasn't worked for a long time :(

**Comment by thaJeztah** [comment]
ping @cpuguy83 PTAL

**Comment by thaJeztah** [comment]
I think this also relates to https://github.com/docker/docker/pull/23436

File: volume/store/store.go [hunk 2]
```diff
@@ -260,7 +260,10 @@ func (s *VolumeStore) create(name, driverName string, opts, labels map[string]st
...
+		if driverName == "" || driverName == volume.DefaultDriverName {
```
Comment (cpuguy83) [inline volume/store/store.go]: Instead of doing this, why not grab the volume driver from the `v` object and do a lookup against the driver?

**Comment by cpuguy83** [comment]
Also, could you make sure to use `git commit -s` to sign your work? For some reason CI does not like the current signature.

**Comment by cpuguy83** [comment]
re: deleting the cache... it's something I started doing, but in reality we do need it, it helps protect against name conflicts, and we need to even enhance it by persisting driver and driver opts into the metadata store (and in turn pulling the metadata store into memory quite likely).

**Comment by LK4D4** [comment]
@moypray could you try to resign your commit? Thanks!

**Comment by erikh** [comment]
This cache, along with nocopy mode vs default, is like 99% of my problem
with docker volume drivers right now. I'd love to explain but the detail is
prohibitively long and arguably off-topic already.

Is there a way we could do a hangout or something with some volumes peeps
to get a better idea of what we need as far as global requirements go? I
don't really think the tracker is the best place to have what could be a
short discussion. I think everyone wins here, including docker.

**Comment by moypray** [comment]
@LK4D4  I am not sure how to fix this issue now.  the local cache is useful. 
 so no matter how to fix it, it will have defect. for example plugin delete a volume which used in a dead containers.... blabla..

File: volume/store/store.go (hunk 2 above)
Comment (cpuguy83) [inline volume/store/store.go]: ping

**Comment by thaJeztah** [comment]
ping @cpuguy83 @LK4D4 ptal!

**Comment by cpuguy83** [comment]
I'm going to carry this, but not ready to make a PR yet.

**Comment by cpuguy83** [comment]
I've carried this in #27164

Thanks!

**Pull request closed without merging**